mllib/src/main/scala/org/apache/spark/mllib/clustering/package.scala

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib
+
+import breeze.linalg.{ Vector => BV }
+
+import org.apache.spark.mllib.linalg.Vector
+import org.apache.spark.rdd.RDD
+
+package object base {
```

Comment:
> 1) modifications to BLAS to support arbitrary functions and axpy on SparseVectors

We already support axpy on SparseVectors in `mllib.linalg.BLAS`. What do you mean by `arbitrary functions`? If the target is to improve k-means, we only need to implement BLAS ops required by k-means.

> 2) definition of PointOps trait and various implementations for different divergences
> 3) rewrite of clusterer to use PointOps trait but only using the squared Euclidean distance implementation
> 4) generalization of clustering interface to select different divergences.

Sounds good.

> 5) store seeds

This is already covered in https://github.com/apache/spark/pull/3610, which is almost ready to merge.

> 6) maintain exactly K clusters by splitting clusters heuristicly.

We had an example in `StreamingKMeans`. But if we want to support different distances, we need to consider how to keep the new centers inside the domain.

Thanks for improving your implementation! Before we merge features one by one, it would be nice to maintain your implementation as a 3rd-party Spark package (spark-packages.org). That helps people find it and send feedbacks.